File: src/common/docblock.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type flow_mode =
  | OptIn
  | OptInStrict
  | OptInStrictLocal
  | OptOut

(*
  * Specifies a function that should be invoked instead of React.createElement
  * when interpreting JSX syntax. Otherwise, the usual rules of JSX are
  * followed: children are varargs after a props argument.
  *)
type jsx_pragma = string * (Loc.t, Loc.t) Flow_ast.Expression.t

type jsx_runtime_pragma =
  | JsxRuntimePragmaClassic
  | JsxRuntimePragmaAutomatic

type t = {
  flow: flow_mode option;
  preventMunge: bool;
  jsx: jsx_pragma option;
  jsxRuntime: jsx_runtime_pragma option;
}

let default_info = { flow = None; preventMunge = false; jsx = None; jsxRuntime = None }

(* accessors *)
let flow info = info.flow

let preventMunge info = info.preventMunge

let jsx info = info.jsx

let jsx_runtime info = info.jsxRuntime

let is_strict info =
  match info.flow with
  | Some OptInStrict -> true
  | Some OptIn
  | Some OptInStrictLocal
  | Some OptOut
  | None ->
    false

let is_flow info =
  match info.flow with
  | Some OptIn
  | Some OptInStrict
  | Some OptInStrictLocal ->
    true
  | Some OptOut
  | None ->
    false

(* debugging *)
let json_of_docblock info =
  Hh_json.(
    let flow =
      match flow info with
      | Some OptIn -> JSON_String "OptIn"
      | Some OptInStrict -> JSON_String "OptInStrict"
      | Some OptInStrictLocal -> JSON_String "OptInStrictLocal"
      | Some OptOut -> JSON_String "OptOut"
      | None -> JSON_Null
    in
    let preventsMunge =
      if preventMunge info then
        JSON_Bool true
      else
        JSON_Null
    in
    JSON_Object [("flow", flow); ("preventMunge", preventsMunge)]
  )

File: rust_port/crates/flow_common/src/docblock.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::sync::Arc;

use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_parser::ast::expression::Expression;
use flow_parser::loc::Loc;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FlowMode {
    OptIn,
    OptInStrict,
    OptInStrictLocal,
    OptOut,
}

#[derive(Debug, Clone)]
pub struct JsxPragma {
    pub raw: String,
    pub expression: Expression<Loc, Loc>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum JsxRuntimePragma {
    Classic,
    Automatic,
}

#[derive(Debug, Clone, Default)]
pub struct Docblock {
    pub flow: Option<FlowMode>,
    pub prevent_munge: bool,
    pub jsx: Option<Arc<JsxPragma>>,
    pub jsx_runtime: Option<JsxRuntimePragma>,
    pub supports_platform: Option<Vec<FlowSmolStr>>,
}

impl Docblock {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn flow(&self) -> Option<FlowMode> {
        self.flow
    }

    pub fn prevent_munge(&self) -> bool {
        self.prevent_munge
    }

    pub fn jsx(&self) -> Option<&JsxPragma> {
        self.jsx.as_deref()
    }

    pub fn jsx_runtime(&self) -> Option<JsxRuntimePragma> {
        self.jsx_runtime
    }

    pub fn is_strict(&self) -> bool {
        matches!(self.flow, Some(FlowMode::OptInStrict))
    }

    pub fn is_flow(&self) -> bool {
        matches!(
            self.flow,
            Some(FlowMode::OptIn | FlowMode::OptInStrict | FlowMode::OptInStrictLocal)
        )
    }

    pub fn supports_platform(&self) -> Option<&[FlowSmolStr]> {
        self.supports_platform.as_deref()
    }
}
